### PR TITLE
[Relay] add fatal

### PR DESCRIPTION
**Pull request by MarisaKirisame**
@jroesch @junrushao1994 @tqchen @wweic @vinx13 @slyubomirsky can you guys review?

### CODE REVIEW

File: python/tvm/relay/expr_functor.py [hunk 1]
```diff
@@ -259,3 +261,6 @@ def visit_ref_write(self, r):
 
     def visit_ref_read(self, r):
         return RefRead(self.visit(r.ref))
+
+    def visit_fatal(self, r):
+        pass
```
Comment (slyubomirsky) [inline python/tvm/relay/expr_functor.py]: Is there a reason this doesn't match the cases for the other nodes?

File: src/relay/pass/type_infer.cc [hunk 2]
```diff
@@ -333,6 +333,10 @@ class TypeInferencer : private ExprFunctor<Type(const Expr&)>,
     return op->op_type;
   }
 
+  Type VisitExpr_(const FatalNode* op) final {
+    return IncompleteTypeNode::make(Kind::kType);
```
Comment (slyubomirsky) [inline src/relay/pass/type_infer.cc]: Is the idea that we want it to be possible to put Fatal inside any subexpression? I was imagining it as a `() -> ()` "statement"

Comment (slyubomirsky) [reply src/relay/pass/type_infer.cc]: (I don't object to the idea of being able to put Fatal anywhere, but we should be clear on the semantics.)

**Comment by slyubomirsky** [comment]
Mostly looks good and it is good to have a more explicit solution to the problem of incomplete matches. A couple of issues besides the questions I've labeled:
* Test cases for hashing and alpha-equal would be good too (they're dull but if anything goes wrong with those, it could be a problem in many areas)
* We'd have to update various interpreters to handle fatal nodes and have test cases for those

File: src/relay/pass/type_infer.cc (hunk 2 above)
Comment (MarisaKirisame) [reply src/relay/pass/type_infer.cc]: it is like haskell's error : a

**Comment by slyubomirsky** [comment]
Thanks for addressing the feedback. I think these changes are good (pending linting fixes)

**Comment by MarisaKirisame** [comment]
@slyubomirsky can you approve explicitly?

File: src/relay/ir/expr.cc [hunk 3]
```diff
@@ -330,6 +330,31 @@ TVM_STATIC_IR_FUNCTOR_REGISTER(IRPrinter, vtable)
...
+  n->type_annotation = std::move(type_annotation);
+  return Fatal(n);
+}
+
+TVM_REGISTER_NODE_TYPE(FatalNode);
+
+TVM_REGISTER_API("relay._make.Fatal")
+.set_body_typed(FatalNode::make);
+
+std::string NoMatchMsg() {
```
Comment (icemelon9) [inline src/relay/ir/expr.cc]: I feel we can just use constexpr or an enum to define some common error messages in both C++ and python. Not necessary to define such an API.
otherwise lgtm

**Comment by MarisaKirisame** [comment]
Hi ppl, I add pytest.ini and updated tests/lint/check_file_type.py, is it ok?

File: src/relay/ir/expr.cc (hunk 3 above)
Comment (MarisaKirisame) [reply src/relay/ir/expr.cc]: I have to copy the string twice if it is a constant though. I feel like api is cleaner and not cumbersome to use too.

**Comment by slyubomirsky** [comment]
Should there be an AST header on the pytest.ini? The linter is complaining about it, but should the .ini file be an exception?

File: tests/python/relay/test_adt.py [hunk 4]
```diff
@@ -684,6 +686,19 @@ def test_iterate():
     assert count(res) == 12
 
 
+def test_head_fatal():
+    a = TypeVar("a")
+    x = Var("x", l(a))
+    y = Var("y")
+    z = Var("z")
+    nil_case = Clause(PatternConstructor(nil, []), Fatal("cannot pass nil into head"))
```
Comment (wweic) [inline tests/python/relay/test_adt.py]: this line is not used.

**Comment by MarisaKirisame** [comment]
@wweic @icemelon9 I had addressed the review comment, can you guys give another round of review or approve?

**Comment by broune** [comment]
I was actually just now looking at support for Asserts, so this will be useful to me (Thanks to Wei for pointing me to this PR). I had been wondering what the TVM Relay policy is on side effects? I see that there is already Ref, but I didn't find the policy on it for Relay, and I wonder what it is. Side effects do change how a lot of optimizations can be done.

(Turns out, if you Google for "TVM side effects", the result is an official message "FDA Warns Public of TVM Side Effects", so I guess there's that, too.)

**Comment by wweic** [comment]
@broune Thanks for bringing it up! cc @tqchen @jroesch @zhiics @icemelon9 

Regarding this PR, @MarisaKirisame please rebase so I can approve.

**Comment by MarisaKirisame** [comment]
@broune relay is a purposefully side-effect language - it has side effect, and it is the key ingridient for our automatic differentiation algorithm.
However, most pass assume the none existence of side effect, because it turn out having ocaml style Ref is very, very hard to reason about! I always planned to add an abstracting abstract machine to relay to deal with them, but it is hard and I have no time. So, they are only half-supported, use at your own risk.

**Comment by tqchen** [comment]
I think it would be good to introduce a thread discussing side effect. My feeling is that we want to keep most of the code functional(which makes them easy to optimize) and only introduce side effect when necessary(e.g. update rules). Side effect was not necessary for first-order AD, and for higher order ADs, we will need to find out a way to reduce as many as possible, and has not yet becomes concern.

In a gist, a large component of optimizations could be focusing on optimizing side-effect free programs. @broune seems to have a lot of thoughts for relaxing the side effect semantics. I think that makes sense especially for asserts, mainly because we only want to receive errors in boundaries of functions(which could be an entire graph) and I think we could reorder things in these cases.

**Comment by jroesch** [comment]
@broune maybe we can have a longer form discussion and draft an RFC together? the current policy is if you don't use ref, the program *should be* side-effect free. I believe this may be violated by a few ops but my goal is to avoid side effects, and make it easy to do side-effect free analysis for many use cases.

**Comment by broune** [comment]
I'm happy to discuss further in whichever venue. (No side effects is ideal for optimization but hard to get 100%. E.g. with dynamic shapes, most ops can have shape errors at runtime, making them side effecting unless somehow the error becomes an output in the dataflow graph. A random number generator library like TVM's tvm.contrib.random has random numbers that are different each time you call the op, which is stateful and a side effect.)

**Comment by MarisaKirisame** [comment]
@tqchen that's the design existence of the PE pass.
However, side effect appear more then in AD.
Dropout introduce randomness, which is a form of side effect - referential transparency is broken. If you copy code that has dropout, they will return different answer when run.
Batchnorm also introduce side effect, as they maintain some global variable.
My take is that, people, sooner or later, has to embrace side effect, for training is full of them.
While that is painful, we can build analysis framework using abstract interpretation to still allow ppl to write basic, first order algorithm (the abstract interpreter will take the responsibility of dealing with control flow, closure, side effect).

on another note, can somebody get this merged? this had been sitting for a long time.

**Comment by wweic** [comment]
@jroesch @zhiics @icemelon9 @tqchen ping.

**Comment by tqchen** [comment]
close for now due to inactive status, feel free to bring another PR to the master
